submit-api/src/Cardano/TxSubmit/Tx.hs

```diff
@@ -1,68 +1,112 @@
+import qualified Cardano.Chain.UTxO as Byron
+import qualified Shelley.Spec.Ledger.BaseTypes as Shelley
+import qualified Shelley.Spec.Ledger.Tx as Shelley
 
-newTxSubmitVar :: IO (TxSubmitVar)
-newTxSubmitVar =
-  TxSubmitVar <$> newEmptyTMVarM <*> newEmptyTMVarM
+-- | An error that can occur while submitting a transaction to a local node.
+data TxSubmitError
+  = TxSubmitByronError !(ApplyTxErr ByronBlock)
+  | TxSubmitShelleyError !(ApplyTxErr (ShelleyBlock TPraosStandardCrypto))
+  | TxSubmitEraMismatchError !EraMismatch
+  deriving (Eq, Show)
 
--- | Read a previously submitted tx from the TMVar.
-readTxSubmit :: TxSubmitVar -> IO (GenTx ByronBlock)
-readTxSubmit tsv =
-  atomically $ takeTMVar (txSubmit tsv)
+renderTxSubmitError :: TxSubmitError -> Text
+renderTxSubmitError tse =
+  case tse of
+    TxSubmitByronError err -> renderApplyMempoolPayloadErr err
+    TxSubmitShelleyError err -> show err -- TODO: Better rendering for Shelley errors
+    TxSubmitEraMismatchError err -> renderEraMismatch err
 
--- | Write the response recieved when tx has been submitted.
-writeTxSubmitResponse
-  :: TxSubmitVar
-  -> (SubmitResult ApplyMempoolPayloadErr)
-  -> IO ()
-writeTxSubmitResponse tsv submitRes =
-  atomically $ putTMVar (txRespond tsv) submitRes
+-- | Submit a transaction to a local node.
+submitTx
+  :: forall mode block.
+     LocalNodeConnectInfo mode block
+  -> (Either (Tx Byron) (Tx Shelley))
+  -> IO (Either TxSubmitError TxId)
+submitTx connectInfo byronOrShelleyTx =
+  case (localNodeConsensusMode connectInfo, byronOrShelleyTx) of
+    (ByronMode{}, Left tx) -> do
+      result <- liftIO $ Api.submitTx connectInfo (TxForByronMode tx)
+      case result of
+        TxSubmitSuccess -> pure $ Right (getTxIdForTx tx)
+        TxSubmitFailureByronMode err ->
+          pure $ Left (TxSubmitByronError err)
 
--- | Submit a tx and wait for the response. This is done as a pair of atomic
--- operations, to allow the tx to be read in one operation, submmited and then
--- the response written as a second operation. Doing this as a single atmomic
--- operation would not work as the other end of the submit/response pair need
--- to be operated on independently.
-submitTx :: TxSubmitVar -> GenTx ByronBlock -> IO (Either TxSubmitError TxId)
-submitTx tsv tx =
-  case tx of
-    ByronTx txid _ -> do
-      atomically $ putTMVar (txSubmit tsv) tx
-      submitRes <- atomically (takeTMVar $ txRespond tsv)
-      case submitRes of
-        SubmitSuccess -> pure $ Right txid
-        SubmitFail r -> pure $ Left (TxSubmitFail r)
-    ByronDlg {} -> pure $ Left $ TxSubmitBadTx "Delegation"
-    ByronUpdateProposal {} -> pure $ Left $ TxSubmitBadTx "Proposal"
-    ByronUpdateVote {} -> pure $ Left $ TxSubmitBadTx "UpdateVote"
+    (ByronMode{}, Right{}) ->
+      pure $ Left $ TxSubmitEraMismatchError EraMismatch {
+                ledgerEraName = "Byron",
+                otherEraName  = "Shelley"
+              }
```

Comment:
Is this where we drop support for Byron only mode?

Reply:
Actually, I haven't removed support for the Byron-only mode, i.e. you can still use it by running the `cardano-submit-api` with the `--byron-mode` flag, if you like. 

In the particular case you've highlighted (`(ByronMode{}, Right{})`), the caller has attempted to submit a Shelley transaction in the Byron mode, which is invalid. Thus, we return a `TxSubmitEraMismatchError`.